stdlib/public/core/NFC.swift

```diff
@@ -0,0 +1,201 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+extension Unicode {
+  internal struct NFC<S: StringProtocol> {
+    let base: S
+  }
+}
+
+extension Unicode.NFC {
+  internal struct Iterator {
+    var buffer: [(scalar: Unicode.Scalar, normData: UInt16)] = []
+    
+    var composee: Unicode.Scalar? = nil
+    
+    var hasBeenReversed = false
+    
+    var iterator: Unicode.NFD<S>.Iterator
+  }
+}
+
+extension Unicode.NFC.Iterator: IteratorProtocol {
+  internal func compose(
+    _ x: Unicode.Scalar,
+    and y: Unicode.Scalar
+  ) -> Unicode.Scalar? {
+    // Fast path: ASCII and some latiny scalars never compose when they're on
+    // the rhs.
+    if y.value < 0x300 {
+      return nil
+    }
+    
+    switch (x.value, y.value) {
+    // Check for Hangul (L, V) -> LV compositions.
+    case (0x1100 ... 0x1112, 0x1161 ... 0x1175):
+      let lIdx = x.value &- 0x1100
+      let vIdx = y.value &- 0x1161
+      let lvIdx = lIdx &* 588 &+ vIdx &* 28
+      let s = 0xAC00 &+ lvIdx
+      return Unicode.Scalar(s)
```

Comment:
Why do some of these use `Unicode.Scalar(_value:)` and others use the failable initializer?